File: libsrc/pylith/sources/RickerFunction.cc
// -*- C++ -*-
//
// ----------------------------------------------------------------------
//
// Brad T. Aagaard, U.S. Geological Survey
// Charles A. Williams, GNS Science
// Matthew G. Knepley, University at Buffalo
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010-2021 University of California, Davis
//
// See LICENSE.md for license information.
//
// ----------------------------------------------------------------------
//

#include <portinfo>

#include "pylith/sources/RickerFunction.hh" // implementation of object methods

#include "pylith/sources/AuxiliaryFactoryRickerFunction.hh" // USES AuxiliaryFactoryRickerFunction
#include "pylith/fekernels/RickerFunction.hh" // USES RickerFunction kernels
#include "pylith/utils/journals.hh" // USES PYLITH_COMPONENT_*
#include "pylith/utils/error.hh" // USES PYLITH_METHOD_BEGIN/END

#include "spatialdata/geocoords/CoordSys.hh" // USES CoordSys

#include <typeinfo> // USES typeid()

// ---------------------------------------------------------------------------------------------------------------------
// Default constructor.
pylith::sources::RickerFunction::RickerFunction(void) :
    _auxiliaryFactory(new pylith::sources::AuxiliaryFactoryRickerFunction)
     {
    pylith::utils::PyreComponent::setName("rickerfunction");
} // constructor


// ---------------------------------------------------------------------------------------------------------------------
// Destructor.
pylith::sources::RickerFunction::~RickerFunction(void) {
    deallocate();
} // destructor


// ---------------------------------------------------------------------------------------------------------------------
// Deallocate PETSc and local data structures.
void
pylith::sources::RickerFunction::deallocate(void) {
    SourceTimeFunctionPointForce::deallocate();

    delete _auxiliaryFactory;_auxiliaryFactory = NULL;
} // deallocate

// ---------------------------------------------------------------------------------------------------------------------
// Get auxiliary factory associated with physics.
pylith::sources::AuxiliaryFactoryPointForce*
pylith::sources::RickerFunction::getAuxiliaryFactory(void) {
    return _auxiliaryFactory;
} // getAuxiliaryFactory


// ---------------------------------------------------------------------------------------------------------------------
// Add source time subfields to auxiliary field.
void
pylith::sources::RickerFunction::addAuxiliarySubfields(void) {
    PYLITH_METHOD_BEGIN;
    PYLITH_COMPONENT_DEBUG("addAuxiliarySubfields(void)");

    // :ATTENTION: The order for adding subfields must match the order of the auxiliary fields in the point-wise
    // functions (kernels).

    _auxiliaryFactory->addRickerCenterFrequency(); // numA - 1
    
    PYLITH_METHOD_END;
} // addAuxiliarySubfields


// ---------------------------------------------------------------------------------------------------------------------
// Get g1v kernel for residual, G(t,s).
PetscPointFunc
pylith::sources::RickerFunction::getKernelg1v_explicit(const spatialdata::geocoords::CoordSys* coordsys) const {
    PYLITH_METHOD_BEGIN;
    PYLITH_COMPONENT_DEBUG("getKernelg1v_explicti(coordsys="<<typeid(coordsys).name()<<")");

    const int spaceDim = coordsys->getSpaceDim();
    PetscPointFunc g1v =
        (3 == spaceDim) ? pylith::fekernels::RickerFunction3D::g1v :
        (2 == spaceDim) ? pylith::fekernels::RickerFunctionPlaneStrain::g1v :
        NULL;

    PYLITH_METHOD_RETURN(g1v);
} // getKernelResidualStress




// End of file

File: pylith/sources/RickerFunction.py
# ----------------------------------------------------------------------
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University at Buffalo
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2021 University of California, Davis
#
# See LICENSE.md for license information.
#
# ----------------------------------------------------------------------
#
# @file pylith/sources/RickerFunction.py
#
# @brief Python source time functiof for a ricker wavelet.
#
# Factory: pointforce_sourcetimefunction

from .SourceTimeFunctionPointForce import SourceTimeFunctionPointForce
from .sources import RickerFunction as ModuleRickerFunction


class RickerFunction(SourceTimeFunctionPointForce, ModuleRickerFunction):
    """Python source time function for ricker source.

    FACTORY: pointforce_sourcetimefunction
    """

    import pythia.pyre.inventory

    # PUBLIC METHODS /////////////////////////////////////////////////////

    def __init__(self, name="rickerfunction"):
        """Constructor.
        """
        SourceTimeFunctionPointForce.__init__(self, name)
        return

    def _defaults(self):
        from .AuxSubfieldsRickerFunction import AuxSubfieldsRickerFunction
        self.auxiliarySubfields = AuxSubfieldsRickerFunction("auxiliary_subfields")

    def preinitialize(self, problem):
        SourceTimeFunctionPointForce.preinitialize(self, problem)


        return

    # PRIVATE METHODS ////////////////////////////////////////////////////

    def _createModuleObj(self):
        """Call constructor for module object for access to C++ object.
        """
        ModuleRickerFunction.__init__(self)


# FACTORIES ////////////////////////////////////////////////////////////

def pointforce_sourcetimefunction():
    """Factory associated with RickerFunction.
    """
    return RickerFunction()


# End of file
